### PR TITLE
Feature/bidder classifications TM-1484

### PR DESCRIPTION
**Pull request by scott062**
Add classifications to front end 

WIP - 
Need to update with designs

Current Issues: 
Labels Names for Classifications
Spacing / Scroll of client badges for classifications
Tests

Would love feedback on any anti-patterns or ideas to avoid drilling the classifications props

### CODE REVIEW

File: src/Components/BidderPortfolio/BidderPortfolioPage/BidderPortfolioPage.jsx [hunk 1]
```diff
@@ -45,8 +46,8 @@ class BidderPortfolioPage extends Component {
       navDataIsLoading = bidderPortfolioCountsIsLoading && !bidderPortfolioCounts.all_clients;
     }
     // for bidder results, however, we'll wait until everything is loaded
-    const bidderPortfolioIsLoadingNotErrored = bidderPortfolioIsLoading &&
-      !bidderPortfolioHasErrored;
+    const bidderPortfolioIsLoadingNotErrored = (bidderPortfolioIsLoading &&
+      !bidderPortfolioHasErrored) && (classificationsIsLoading && !classificationsHasErrored);
```
Comment (mjoyce91) [inline src/Components/BidderPortfolio/BidderPortfolioPage/BidderPortfolioPage.jsx]: The extra parens don't do any thing here

File: src/Components/BidderPortfolio/BidderPortfolioStatCard/BidderPortfolioStatCard.jsx [hunk 2]
```diff
@@ -2,18 +2,18 @@ import React from 'react';
...
 import ClientBadgeList from '../ClientBadgeList';
-import StaticDevContent from '../../StaticDevContent';
 import SearchAsClientButton from '../SearchAsClientButton';
 
 const useCDOBidding = () => checkFlag('flags.cdo_bidding');
 
-const BidderPortfolioStatCard = ({ userProfile }) => {
+const BidderPortfolioStatCard = ({ userProfile, classifications }) => {
   const currentAssignmentText = get(userProfile, 'pos_location_code');
+  const clientClassifications = userProfile.classifications;
```
Comment (mjoyce91) [inline src/Components/BidderPortfolio/BidderPortfolioStatCard/BidderPortfolioStatCard.jsx]: Let's do `get(userProfile, 'classifications', [])`

File: src/Components/BidderPortfolio/CheckboxList/CheckboxList.jsx [hunk 3]
```diff
@@ -1,32 +1,34 @@
...
         small
-        value
-        key={c.value}
-        disabled={isDisabled}
-        className={isDisabled ? 'tm-checkbox-disabled-alternate' : ''}
+        value={clientClassifications.indexOf(c.code) > -1}
+        key={c.code}
+        disabled
+        checked={clientClassifications.indexOf(c.code) > -1}
+        className={'tm-checkbox-disabled-alternate'}
```
Comment (mjoyce91) [inline src/Components/BidderPortfolio/CheckboxList/CheckboxList.jsx]: `className="tm-checkbox-disabled-alternate"`

File: src/Components/BidderPortfolio/CheckboxList/CheckboxList.jsx [hunk 4]
```diff
@@ -1,32 +1,34 @@
...
+        label={c.text}
         small
-        value
-        key={c.value}
-        disabled={isDisabled}
-        className={isDisabled ? 'tm-checkbox-disabled-alternate' : ''}
+        value={clientClassifications.indexOf(c.code) > -1}
+        key={c.code}
+        disabled
+        checked={clientClassifications.indexOf(c.code) > -1}
```
Comment (mjoyce91) [inline src/Components/BidderPortfolio/CheckboxList/CheckboxList.jsx]: Let's move this logic to the top of the map:

```jsx
{list.map(c => {
  const checked = indexOf(clientClassifications, c.code) > -1; // use lodash indexOf here
  ...
  value={checked}
  checked={checked}
  ...
```

File: src/Components/BidderPortfolio/ClientBadge/ClientBadge.jsx [hunk 5]
```diff
@@ -3,33 +3,89 @@ import PropTypes from 'prop-types';
 import FA from 'react-fontawesome';
 
 const icons = {
-  handshake: {
+  3: {
     isIcon: true,
-    name: 'handshake-o',
-    text: 'Handshake',
+    name: 'bath',
```
Comment (mjoyce91) [inline src/Components/BidderPortfolio/ClientBadge/ClientBadge.jsx]: 🛀

File: src/Components/BidderPortfolio/ClientBadge/ClientBadge.jsx [hunk 6]
```diff
@@ -3,33 +3,89 @@ import PropTypes from 'prop-types';
...
+    text: 'Recommended for Tenure',
+  },
+  6: {
     name: '6/8',
-    text: '6/8',
+    text: '6/8 Rule',
+  },
+  A: {
+    isIcon: true,
+    name: 'ship',
```
Comment (mjoyce91) [inline src/Components/BidderPortfolio/ClientBadge/ClientBadge.jsx]: 🚢

File: src/Components/BidderPortfolio/ClientBadge/ClientBadge.jsx [hunk 7]
```diff
@@ -3,33 +3,89 @@ import PropTypes from 'prop-types';
...
-  fairshare: {
+  F: {
     isIcon: true,
     name: 'balance-scale',
     text: 'Fair Share',
   },
-  retirement: {
+  F1: {
+    isIcon: true,
+    name: 'graduation-cap',
```
Comment (mjoyce91) [inline src/Components/BidderPortfolio/ClientBadge/ClientBadge.jsx]: 👨‍🎓

File: src/Components/BidderPortfolio/ClientBadge/ClientBadge.jsx [hunk 8]
```diff
@@ -3,33 +3,89 @@ import PropTypes from 'prop-types';
...
+    name: 'users',
+    text: 'Tandem',
   },
 };
 
-const status$ = ['none', 'warning', 'success'];
+const status$ = ['none', 'success'];
 
 const ClientBadge = ({ type, status }) => {
+  const isHighlighted = (status === true) ? 'success' : 'none';
```
Comment (mjoyce91) [inline src/Components/BidderPortfolio/ClientBadge/ClientBadge.jsx]: don't need parens

File: src/Components/BidderPortfolio/ClientBadge/ClientBadge.jsx [hunk 9]
```diff
@@ -49,6 +105,27 @@ ClientBadge.propTypes = {
   status: PropTypes.oneOf([0, 1, 2]),
 };
 
+ClientBadge.propTypes = {
+  type: PropTypes.oneOf([
```
Comment (mjoyce91) [inline src/Components/BidderPortfolio/ClientBadge/ClientBadge.jsx]: 👍

File: src/Components/BidderPortfolio/ClientBadge/ClientBadge.jsx [hunk 10]
```diff
@@ -49,6 +105,27 @@ ClientBadge.propTypes = {
...
+    'differential',
+    'fairshare',
+    'pickering',
+    'rangel',
+    'meritorious',
+    'fellow',
+    'tenure',
+    'tandem',
+  ]).isRequired,
+  status: PropTypes.oneOf([0, 1, 2]),
```
Comment (mjoyce91) [inline src/Components/BidderPortfolio/ClientBadge/ClientBadge.jsx]: 👍

File: src/Components/BidderPortfolio/ClientBadgeList/ClientBadgeList.jsx [hunk 11]
```diff
@@ -1,27 +1,26 @@
...
+const ClientBadgeList = ({ classifications, clientClassifications }) => (
   <div className="usa-grid-full client-badge-list">
-    <ClientBadge type="handshake" status={statuses.handshake} />
-    <ClientBadge type="sixeight" status={statuses.sixeight} />
-    <ClientBadge type="fairshare" status={statuses.fairshare} />
-    <ClientBadge type="retirement" status={statuses.retirement} />
+    {classifications.map(c => (
+      <ClientBadge
+        type={c.code}
+        status={clientClassifications.indexOf(c.code) > -1}
```
Comment (mjoyce91) [inline src/Components/BidderPortfolio/ClientBadgeList/ClientBadgeList.jsx]: Same as https://github.com/MetaPhase-Consulting/State-TalentMAP/pull/727/files#r383901047

File: src/Components/ProfileDashboard/Classifications/Classifications.jsx [hunk 12]
```diff
@@ -0,0 +1,44 @@
+import React, { Component } from 'react';
+import SectionTitle from '../SectionTitle';
+import CheckboxList from '../../BidderPortfolio/CheckboxList';
+import { CLASSIFICATIONS, CLIENT_CLASSIFICATIONS } from '../../../Constants/PropTypes';
+
+// eslint-disable-next-line react/prefer-stateless-function
+class Classifications extends Component {
+  // eslint-disable-next-line no-useless-constructor
```
Comment (mjoyce91) [inline src/Components/ProfileDashboard/Classifications/Classifications.jsx]: these disables still needed?

File: src/Containers/ProfilePublic/ProfilePublic.jsx [hunk 13]
```diff
@@ -4,31 +4,47 @@ import { connect } from 'react-redux';
...
 import { push } from 'react-router-redux';
 import { get } from 'lodash';
+import ProfileDashboard from 'Components/ProfileDashboard';
+import { fetchClassifications } from 'actions/classifications';
+import { userProfilePublicFetchData } from 'actions/userProfilePublic';
 import { USER_PROFILE, EMPTY_FUNCTION } from '../../Constants/PropTypes';
 import { DEFAULT_USER_PROFILE } from '../../Constants/DefaultProps';
-import ProfileDashboard from '../../Components/ProfileDashboard';
-import { userProfilePublicFetchData } from '../../actions/userProfilePublic';
+
```
Comment (mjoyce91) [inline src/Containers/ProfilePublic/ProfilePublic.jsx]: extra line break needed?

File: src/Containers/ProfilePublic/ProfilePublic.jsx [hunk 14]
```diff
@@ -4,31 +4,47 @@ import { connect } from 'react-redux';
...
 import Alert from '../../Components/Alert';
 
 class ProfilePublic extends Component {
 
   componentWillMount() {
-    this.props.fetchData(get(this.props, 'match.params.id'));
+    const id = get(this.props, 'match.params.id');
+    this.props.fetchData(id);
```
Comment (mjoyce91) [inline src/Containers/ProfilePublic/ProfilePublic.jsx]: 👍

File: src/actions/classifications.js [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+export function fetchClassifications() {
+  return (dispatch) => {
+    dispatch(classificationsIsLoading(true));
+    dispatch(classificationsHasErrored(false));
+
+    api()
+      .get('/fsbid/reference/classifications/')
+      .then(({ data }) => {
+        dispatch(classificationsHasErrored(false));
+        dispatch(classificationsIsLoading(false));
```
Comment (mjoyce91) [inline src/actions/classifications.js]: Set isLoading as the last dispatch. Otherwise, the component might try to render before `success` has been set.

**Comment by mjoyce91** [comment]
This looks good. I wonder if we need to add a key at the top to denote that the check means the user has the classification. Otherwise, if the user has no classifications, it might look like they actually have all of them.
Screen Shot 2020-02-25 at 9 22 38 AM

**Comment by mjoyce91** [comment]
> Would love feedback on any anti-patterns or ideas to avoid drilling the classifications props

I wouldn't be opposed to making `ClientBadgeList` a connected component

File: src/Components/BidderPortfolio/BidderPortfolioPage/BidderPortfolioPage.jsx (hunk 1 above)
Comment (scott062) [reply src/Components/BidderPortfolio/BidderPortfolioPage/BidderPortfolioPage.jsx]: I'm just really extra 😃

File: src/Components/BidderPortfolio/BidderPortfolioStatCard/BidderPortfolioStatCard.jsx (hunk 2 above)
Comment (scott062) [reply src/Components/BidderPortfolio/BidderPortfolioStatCard/BidderPortfolioStatCard.jsx]: fixed in both stat card and grid

File: src/Components/BidderPortfolio/CheckboxList/CheckboxList.jsx (hunk 4 above)
Comment (scott062) [reply src/Components/BidderPortfolio/CheckboxList/CheckboxList.jsx]: 👍

File: src/Components/BidderPortfolio/CheckboxList/CheckboxList.jsx (hunk 3 above)
Comment (scott062) [reply src/Components/BidderPortfolio/CheckboxList/CheckboxList.jsx]: ✅

File: src/Components/BidderPortfolio/ClientBadge/ClientBadge.jsx (hunk 8 above)
Comment (scott062) [reply src/Components/BidderPortfolio/ClientBadge/ClientBadge.jsx]: (ok)

File: src/Components/BidderPortfolio/ClientBadge/ClientBadge.jsx (hunk 6 above)
Comment (scott062) [reply src/Components/BidderPortfolio/ClientBadge/ClientBadge.jsx]: testing not resolved but commented on

File: src/Components/BidderPortfolio/ClientBadge/ClientBadge.jsx (hunk 7 above)
Comment (scott062) [reply src/Components/BidderPortfolio/ClientBadge/ClientBadge.jsx]: testing resolved and commented on

File: src/Components/BidderPortfolio/ClientBadge/ClientBadge.jsx (hunk 8 above)
Comment (scott062) [reply src/Components/BidderPortfolio/ClientBadge/ClientBadge.jsx]: testing - code changed and commented but not resolved

**Comment by mjoyce91** [comment]
@scott062 Also need to merge latest dev and resolve conflicts

**Comment by mjoyce91** [comment]
@scott062 still some merge conflicts

File: src/Components/ProfileDashboard/Classifications/Classifications.jsx [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+              list={classifications}
+              clientClassifications={clientClassifications}
+              id="updates"
+            />
+          </div>
+        </div>
+      </div>
+    );
+  }
+}
```
Comment (mjoyce91) [inline src/Components/ProfileDashboard/Classifications/Classifications.jsx]: This component still renders even if `classifications` has been fetched, but `clientClassifications` has not been fetched. This results in a small amount of time between when the component renders and shows everything unchecked:

Screen Shot 2020-02-26 at 9 03 37 AM

Screen Shot 2020-02-26 at 9 04 58 AM

File: src/Containers/ProfilePublic/ProfilePublic.jsx [hunk 17]
```diff
@@ -4,31 +4,46 @@ import { connect } from 'react-redux';
...
+      hasErrored,
+      userProfile,
+      classifications,
+      classificationsIsLoading,
+      classificationsHasErrored,
+    } = this.props;
     const { assignments, bidList } = userProfile;
+    const clientClassifications = userProfile.classifications;
     const assignment = get(assignments, '[0]');
+    const combinedLoading = isLoading && classificationsIsLoading;
```
Comment (mjoyce91) [inline src/Containers/ProfilePublic/ProfilePublic.jsx]: This should be an or `||` and will fix https://github.com/MetaPhase-Consulting/State-TalentMAP/pull/727/files#r384512201

File: src/Containers/ProfilePublic/ProfilePublic.jsx [hunk 18]
```diff
@@ -4,31 +4,46 @@ import { connect } from 'react-redux';
...
+    const combinedErrored = hasErrored && classificationsHasErrored;
```
Comment (mjoyce91) [inline src/Containers/ProfilePublic/ProfilePublic.jsx]: This should also an or `||`. Basically with isLoading/hasErrored-type states, you want them to be less restrictive

File: src/Components/ProfileDashboard/Classifications/Classifications.test.jsx [hunk 19]
```diff
@@ -1,29 +1,29 @@
 import React from 'react';
 import { shallow } from 'enzyme';
 import toJSON from 'enzyme-to-json';
-import Updates from './Updates';
+import Classifications from './Classifications';
```
Comment (mjoyce91) [inline src/Components/ProfileDashboard/Classifications/Classifications.test.jsx]: 👍

File: src/Components/ProfileDashboard/Classifications/Classifications.jsx (hunk 12 above)
Comment (scott062) [reply src/Components/ProfileDashboard/Classifications/Classifications.jsx]: // disable the eslint-disable 👍

File: src/Containers/ProfilePublic/ProfilePublic.jsx (hunk 17 above)
Comment (scott062) [reply src/Containers/ProfilePublic/ProfilePublic.jsx]: 🎣 Thanks for the 'good catch'!

**Comment by mjoyce91** [comment]
@scott062 I think most of the failing tests are snapshots

**Comment by mjoyce91** [comment]
@scott062 take a look at the screen cap below. I'm noticing two things:
- The spinner does not display on the Bidder Portfolio when results are loading
- The classifications icons on the Bidder Portfolio change during the loading process

Feb-27-2020 12-16-21

File: src/Components/BidderPortfolio/BidderPortfolioPage/BidderPortfolioPage.jsx [hunk 20]
```diff
@@ -49,7 +49,7 @@ class BidderPortfolioPage extends Component {
     }
     // for bidder results, however, we'll wait until everything is loaded
     const bidderPortfolioIsLoadingNotErrored = bidderPortfolioIsLoading &&
-      !bidderPortfolioHasErrored;
+      !bidderPortfolioHasErrored && classificationsIsLoading && !classificationsHasErrored;
```
Comment (mjoyce91) [inline src/Components/BidderPortfolio/BidderPortfolioPage/BidderPortfolioPage.jsx]: Ah, my screenshot is related to this. There's a couple things going on here:

This should be changed to:
```jsx
const bidderPortfolioIsLoadingNotErrored =
    (bidderPortfolioIsLoading || classificationsIsLoading) &&
    !bidderPortfolioHasErrored && !classificationsHasErrored;
```

You also seem to be passing the function, not the result, of `classificationsIsLoading` and `classificationsHasErrored`

File: src/Containers/BidderPortfolio/BidderPortfolio.jsx [hunk 21]
```diff
@@ -4,10 +4,11 @@ import { get, isEqual, pick } from 'lodash';
 import queryString from 'query-string';
 import { connect } from 'react-redux';
 import { withRouter } from 'react-router';
+import { fetchClassifications, classificationsIsLoading, classificationsHasErrored } from 'actions/classifications';
```
Comment (mjoyce91) [inline src/Containers/BidderPortfolio/BidderPortfolio.jsx]: You don't need to import the isLoading and hasErrored, only the fetch. isLoading and hasErrored are captured in mapStateToProps

File: src/Containers/BidderPortfolio/BidderPortfolio.jsx [hunk 22]
```diff
@@ -128,6 +131,9 @@ class BidderPortfolio extends Component {
         bidderPortfolioCounts={bidderPortfolioCounts}
         bidderPortfolioCountsIsLoading={bidderPortfolioCountsIsLoading}
         bidderPortfolioCountsHasErrored={bidderPortfolioCountsHasErrored}
+        classificationsIsLoading={classificationsIsLoading}
```
Comment (mjoyce91) [inline src/Containers/BidderPortfolio/BidderPortfolio.jsx]: This is referencing that import above, not your prop

File: src/sass/_profile.scss [hunk 23]
```diff
@@ -99,11 +99,17 @@
       flex-wrap: wrap;
     }
 
+    .client-checkbox-list > :first-child {
+      border: 0;
+      border-top: 1px solid #AEB0B5;
+      border-bottom: 1px solid #AEB0B5;
```
Comment (mjoyce91) [inline src/sass/_profile.scss]: See if these variables are in the _variable.scss file, if not, either use an existing color from there or add this one

File: src/Containers/BidderPortfolio/BidderPortfolio.jsx (hunk 22 above)
Comment (scott062) [reply src/Containers/BidderPortfolio/BidderPortfolio.jsx]: I want this to come from the state that i've mapped correct?

File: src/Containers/BidderPortfolio/BidderPortfolio.jsx (hunk 21 above)
Comment (scott062) [reply src/Containers/BidderPortfolio/BidderPortfolio.jsx]: 👍

File: src/sass/_profile.scss (hunk 23 above)
Comment (scott062) [reply src/sass/_profile.scss]: $bg-gray-dark-2 🐺

File: src/Containers/BidderPortfolio/BidderPortfolio.jsx (hunk 22 above)
Comment (mjoyce91) [reply src/Containers/BidderPortfolio/BidderPortfolio.jsx]: correct, which you can access via `this.props`

**Comment by mjoyce91** [comment]
@scott062 check failing tests

File: src/Components/BidderPortfolio/ClientBadgeList/ClientBadgeList.jsx [hunk 24]
```diff
@@ -1,27 +1,35 @@
...
+      .map((c) => {
+        const checked = clientClassifications.includes(c.code);
+        return (
+          <ClientBadge
+            type={c.code}
+            status={checked}
+          />
+        );
+      })
+    }
```
Comment (scott062) [inline src/Components/BidderPortfolio/ClientBadgeList/ClientBadgeList.jsx]: I think this is why the order of those badges changes. Its sorting only into two groups of either checked or unchecked, but it doesn't consistently display those two subsets.

File: src/Components/BidderPortfolio/ClientBadgeList/ClientBadgeList.jsx [hunk 25]
```diff
@@ -1,18 +1,17 @@
...
 import ClientBadge from '../ClientBadge';
 import { CLASSIFICATIONS, CLIENT_CLASSIFICATIONS } from '../../../Constants/PropTypes';
 
 const ClientBadgeList = ({ classifications, clientClassifications }) => (
   <div className="usa-grid-full client-badge-list">
-    {classifications.sort((c) => {
-      const checked = clientClassifications.includes(c.code);
-      return (checked ? -1 : 1);
-    }).slice(0, 4)
+    {orderBy(classifications, c => clientClassifications.includes(c.code), ['desc'])
```
Comment (mjoyce91) [inline src/Components/BidderPortfolio/ClientBadgeList/ClientBadgeList.jsx]: 👍

**Pull request merged**